File: apps/idp-core/app/catalog.py
import json
from pathlib import Path
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class CatalogModel(BaseModel):
    model_config = ConfigDict(extra="allow")


class EnvironmentSpec(CatalogModel):
    name: str | None = None
    type: str | None = None
    namespace: str | None = None
    route: str | None = None
    rbac: dict[str, Any] = Field(default_factory=dict)
    deployment: dict[str, Any] = Field(default_factory=dict)
    health: Any = None
    sync: Any = None


class SecretSpec(CatalogModel):
    name: str | None = None
    binding: str = "unknown"
    rotation: str = "unknown"


class ScorecardSpec(CatalogModel):
    runtime_profile: str = "unknown"
    has_health_endpoint: bool = False
    has_network_policy: bool = False
    has_owner: bool = False


class ApplicationSpec(CatalogModel):
    name: str | None = None
    display_name: str | None = None
    owner: str | None = None
    lifecycle: str | None = None
    source: dict[str, Any] = Field(default_factory=dict)
    deployment: dict[str, Any] = Field(default_factory=dict)
    environments: list[EnvironmentSpec] = Field(default_factory=list)
    secrets: list[SecretSpec] = Field(default_factory=list)
    scorecard: ScorecardSpec = Field(default_factory=ScorecardSpec)
    health: Any = None


class CatalogDocument(CatalogModel):
    schema_version: str | None = None
    core_components: dict[str, Any] = Field(default_factory=dict)
    applications: list[ApplicationSpec] = Field(default_factory=list)


class DeploymentRecord(CatalogModel):
    app: str | None = None
    environment: str | None = None
    route: str | None = None
    controller: Any = None
    image: str | None = None
    health: str | None = None
    sync: str | None = None


class SecretBinding(CatalogModel):
    app: str | None = None
    name: str | None = None
    binding: str = "unknown"
    rotation: str = "unknown"


class ScorecardRecord(CatalogModel):
    app: str | None = None
    runtime_profile: str = "unknown"
    has_health_endpoint: bool = False
    has_network_policy: bool = False
    has_owner: bool = False


class ApplicationSurfaceRecord(CatalogModel):
    app: str | None = None
    display_name: str | None = None
    owner: str | None = None
    lifecycle: str | None = None
    environment: str | None = None
    environment_type: str | None = None
    namespace: str | None = None
    route: str | None = None
    rbac_group: str | None = None
    source_path: str | None = None
    kubernetes_label_selector: str | None = None


def load_catalog(path: Path) -> CatalogDocument:
    return CatalogDocument.model_validate(json.loads(path.read_text(encoding="utf-8")))


def application_specs(catalog: CatalogDocument) -> list[ApplicationSpec]:
    return catalog.applications


def find_application(catalog: CatalogDocument, app_name: str) -> ApplicationSpec | None:
    for app_spec in application_specs(catalog):
        if app_spec.name == app_name:
            return app_spec
    return None


def deployment_records(catalog: CatalogDocument) -> list[DeploymentRecord]:
    records: list[DeploymentRecord] = []
    for app_spec in application_specs(catalog):
        deployment = app_spec.deployment
        for environment in app_spec.environments:
            environment_deployment = environment.deployment
            records.append(
                DeploymentRecord(
                    app=app_spec.name,
                    environment=environment.name,
                    route=environment.route,
                    controller=deployment.get("controller"),
                    image=_optional_string(environment_deployment.get("image") or deployment.get("image")),
                    health=_optional_string(environment.health or app_spec.health),
                    sync=_optional_string(environment.sync or deployment.get("sync")),
                )
            )
    return records


def secret_bindings(catalog: CatalogDocument) -> list[SecretBinding]:
    records: list[SecretBinding] = []
    for app_spec in application_specs(catalog):
        for secret in app_spec.secrets:
            records.append(SecretBinding.model_validate({"app": app_spec.name, **secret.model_dump()}))
    return records


def scorecard_records(catalog: CatalogDocument) -> list[ScorecardRecord]:
    records: list[ScorecardRecord] = []
    for app_spec in application_specs(catalog):
        scorecard = app_spec.scorecard
        scorecard_payload = scorecard.model_dump()
        if "has_owner" not in scorecard.model_fields_set:
            scorecard_payload["has_owner"] = bool(app_spec.owner)
        records.append(
            ScorecardRecord.model_validate(
                {
                    "app": app_spec.name,
                    **scorecard_payload,
                }
            )
        )
    return records


def application_surface_records(catalog: CatalogDocument) -> list[ApplicationSurfaceRecord]:
    records: list[ApplicationSurfaceRecord] = []
    for app_spec in application_specs(catalog):
        for environment in app_spec.environments:
            if not environment.route:
                continue
            records.append(
                ApplicationSurfaceRecord(
                    app=app_spec.name,
                    display_name=app_spec.display_name,
                    owner=app_spec.owner,
                    lifecycle=app_spec.lifecycle,
                    environment=environment.name,
                    environment_type=environment.type,
                    namespace=environment.namespace,
                    route=environment.route,
                    rbac_group=_optional_string(environment.rbac.get("group")),
                    source_path=_optional_string(app_spec.source.get("path")),
                    kubernetes_label_selector=f"app={app_spec.name}" if app_spec.name else None,
                )
            )
    return records


def _optional_string(value: Any) -> str | None:
    return value if isinstance(value, str) else None

File: apps/idp-core/app/internal/catalog/catalog.go
package catalog

import (
	"encoding/json"
	"fmt"
	"os"
)

type Catalog struct {
	SchemaVersion  string            `json:"schema_version"`
	CoreComponents map[string]string `json:"core_components"`
	Applications   []Application     `json:"applications"`
}

type Application struct {
	Name         string        `json:"name"`
	DisplayName  string        `json:"display_name"`
	Owner        string        `json:"owner"`
	Lifecycle    string        `json:"lifecycle"`
	Source       Source        `json:"source"`
	Deployment   Deployment    `json:"deployment"`
	Environments []Environment `json:"environments"`
	Secrets      []Secret      `json:"secrets"`
	Scorecard    Scorecard     `json:"scorecard"`
	Health       string        `json:"health,omitempty"`
}

type Source struct {
	Repo string `json:"repo"`
	Path string `json:"path"`
}

type Deployment struct {
	Strategy     string   `json:"strategy"`
	Controller   string   `json:"controller"`
	Applications []string `json:"applications"`
	Image        string   `json:"image,omitempty"`
	Sync         string   `json:"sync,omitempty"`
}

type Environment struct {
	Name       string      `json:"name"`
	Type       string      `json:"type"`
	Namespace  string      `json:"namespace"`
	Route      string      `json:"route"`
	RBAC       RBAC        `json:"rbac"`
	Deployment *Deployment `json:"deployment,omitempty"`
	Health     string      `json:"health,omitempty"`
	Sync       string      `json:"sync,omitempty"`
}

type RBAC struct {
	Group   string `json:"group"`
	Viewer  bool   `json:"viewer"`
	Mutator bool   `json:"mutator"`
}

type Secret struct {
	Name     string `json:"name"`
	Binding  string `json:"binding"`
	Rotation string `json:"rotation"`
	Scope    string `json:"scope,omitempty"`
}

type Scorecard struct {
	RuntimeProfile            string `json:"runtime_profile"`
	HasHealthEndpoint         bool   `json:"has_health_endpoint"`
	HasNetworkPolicy          bool   `json:"has_network_policy"`
	HasRuntimeProbes          bool   `json:"has_runtime_probes"`
	HasResourceRequestsLimits bool   `json:"has_resource_requests_limits"`
	HasOwner                  bool   `json:"has_owner"`
	HasModelCard              bool   `json:"has_model_card,omitempty"`
	Tier                      string `json:"tier,omitempty"`
}

func Load(path string) (*Catalog, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read catalog file: %w", err)
	}
	var c Catalog
	if err := json.Unmarshal(data, &c); err != nil {
		return nil, fmt.Errorf("unmarshal catalog: %w", err)
	}
	return &c, nil
}

func (c *Catalog) GetApp(name string) (*Application, bool) {
	for _, app := range c.Applications {
		if app.Name == name {
			return &app, true
		}
	}
	return nil, false
}

type DeploymentRecord struct {
	App         string `json:"app"`
	Environment string `json:"environment"`
	Route       any    `json:"route"`
	Controller  string `json:"controller"`
	Image       string `json:"image"`
	Health      string `json:"health"`
	Sync        string `json:"sync"`
}

func (c *Catalog) ListDeployments() []DeploymentRecord {
	var records []DeploymentRecord
	for _, app := range c.Applications {
		for _, env := range app.Environments {
			image := app.Deployment.Image
			if env.Deployment != nil && env.Deployment.Image != "" {
				image = env.Deployment.Image
			}

			health := app.Health
			if env.Health != "" {
				health = env.Health
			}

			sync := app.Deployment.Sync
			if env.Sync != "" {
				sync = env.Sync
			}

			var route any = env.Route
			if env.Route == "" {
				route = nil
			}

			records = append(records, DeploymentRecord{
				App:         app.Name,
				Environment: env.Name,
				Route:       route,
				Controller:  app.Deployment.Controller,
				Image:       image,
				Health:      health,
				Sync:        sync,
			})
		}
	}
	return records
}

type SecretRecord struct {
	Name     string `json:"name"`
	Binding  string `json:"binding"`
	Rotation string `json:"rotation"`
	App      string `json:"app"`
	Scope    string `json:"scope,omitempty"`
}

func (c *Catalog) ListSecrets() []SecretRecord {
	var records []SecretRecord
	for _, app := range c.Applications {
		for _, s := range app.Secrets {
			binding := s.Binding
			if binding == "" {
				binding = "not declared"
			}
			rotation := s.Rotation
			if rotation == "" {
				rotation = "not declared"
			}
			records = append(records, SecretRecord{
				Name:     s.Name,
				Binding:  binding,
				Rotation: rotation,
				App:      app.Name,
				Scope:    s.Scope,
			})
		}
	}
	return records
}

type ScorecardRecord struct {
	App                       string `json:"app"`
	RuntimeProfile            string `json:"runtime_profile"`
	HasHealthEndpoint         bool   `json:"has_health_endpoint"`
	HasNetworkPolicy          bool   `json:"has_network_policy"`
	HasRuntimeProbes          bool   `json:"has_runtime_probes"`
	HasResourceRequestsLimits bool   `json:"has_resource_requests_limits"`
	HasOwner                  bool   `json:"has_owner"`
	HasModelCard              *bool  `json:"has_model_card,omitempty"`
	Tier                      string `json:"tier,omitempty"`
}

func (c *Catalog) ListScorecards() []ScorecardRecord {
	var records []ScorecardRecord
	for _, app := range c.Applications {
		profile := app.Scorecard.RuntimeProfile
		if profile == "" {
			profile = "not declared"
		}

		hasOwner := app.Scorecard.HasOwner
		if app.Owner != "" {
			hasOwner = true
		}

		record := ScorecardRecord{
			App:                       app.Name,
			RuntimeProfile:            profile,
			HasHealthEndpoint:         app.Scorecard.HasHealthEndpoint,
			HasNetworkPolicy:          app.Scorecard.HasNetworkPolicy,
			HasRuntimeProbes:          app.Scorecard.HasRuntimeProbes,
			HasResourceRequestsLimits: app.Scorecard.HasResourceRequestsLimits,
			HasOwner:                  hasOwner,
			Tier:                      app.Scorecard.Tier,
		}

		// Optional field
		if app.Scorecard.HasModelCard {
			val := true
			record.HasModelCard = &val
		}

		records = append(records, record)
	}
	return records
}
